docs: add README

colhom opened the pull request:
addresses https://github.com/coreos/khealth/issues/9

colhom committed:
docs: add README

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+khealth
```
Atx headers: `# khealth`

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+
+[![Docker Repository on Quay.io](https://quay.io/repository/coreos/khealth/status "Docker Repository on Quay.io")](https://quay.io/repository/coreos/khealth)
+
+**khealth** is a kubernetes cluster monitoring suite. Routines are defined which exercise various kubernetes subsystems. Collectors receive status update events from Routines and compute current state.
```
Usually just _italics_ for introduction of new term of art, command name, etc.

joshix replied:
_khealth_ is a Kubernetes cluster monitoring suite. Its Routines exercise Kubernetes subsystems and send events to Collectors. Collectors collate these events to compute current cluster state. Cluster status is available from Collectors over a simple HTTP API, which is served on a cluster nodeport in the example below. 

(That last sentence or something like it -- e.g., a line that says what it does, how one uses or views it.)

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+khealth
+=====
+
+[![Docker Repository on Quay.io](https://quay.io/repository/coreos/khealth/status "Docker Repository on Quay.io")](https://quay.io/repository/coreos/khealth)
```
"Docker Image on Quay.io" might be clearer than Repository.

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+  The program exposes a single health endpoint which reports the state of the latest event.
+
+## Roadmap
+- More routines: We want routines that do everything! Test network latency. Write to disk. Computer fibonacci sequences.
```
s/Computer/Compute/

(and doesn't one just compute _the_ Lamé sequence?)

joshix replied:
Most of the docs use the `* Item` unordered list style.

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+
+- **pkg/routines**:
+
+  Routines are defined via structs that implement the handler interface.
```
Routines are defined in structures that implement the `RoutineHandler` interface.

(Soft suggestion; this just seems a kind of common expression.)

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+	Poll() error
+	Cleanup() error
+  }
+  Init() is called, and then Poll() in a loop. Once the ttl has expired, cleanup is called. Rinse-repeat. Every step generates an event which is received by the collector.
```
This sentence belongs outside the closing three ticks

joshix replied:
If it is `Init()` and `Poll()`, then it should be `Cleanup()`, too.

(or better yet `Init`, `Poll`, and `Cleanup`. I don't think the argument parens are needed in documentation sentences.)

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+
+- **Cluster administrators**: Gain insight into your kubernetes cluster's performance.  Monitor health endpoints which report on various testing routines.
+
+- **Kubernetes developers**: A convenient way to "work-out" a cluster. Feel free to write modules that exist solely to torture-test a cluster and have no business running on the same cluster as production assests. And turn the replica count way up!
```
- "smoke test" maybe rather than "work-out"? (soft suggest)
- "torture test" doesn't need a hyphen.
- `s/assests/assets/`

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+## Roadmap
+- More routines: We want routines that do everything! Test network latency. Write to disk. Computer fibonacci sequences.
+
+- Prometheus integration: collectors expose prometheus-compatbile statyus endpoints and metrics. Pre-built infrastructure for aggregating metrics from a dynamic set of canary pods.
```
- `s/collectors/Collectors/`
- `s/compatbile statyus/compatible status/`
- maybe very briefly and inline explain "canary" in this context? (eg, maybe say something like "...from a dynamic set of "canary pods" purpose-built to exercise Kubernetes clusters")

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+
+## Architecture
+
+A module is defined as a single entrypoint command which makes use of a set of routines and a single collector to expose a metric set via the collector's status endpoint.
```
A khealth _Module_ is a single command that invokes a set of Routines and a single Collector. The Collector gathers events from the Routines and exposes metrics on its status endpoint.

joshix commented on README.md:
```diff
@@ -0,0 +1,74 @@
+
+- **cmd/**:
+
+  This is where the module entrypoint programs live. Each module should have exactly one package in cmd/ which produces an executable which "enacts" that module.
```
- `s/module/Module/g`
- "...exactly one `main` package..."? (then you could cut the sentence to: "Each Module should have exactly one `main` package in an eponymous directory beneath `cmd/`." (because `main` implies an executable command, right?)

colhom committed:
docs: apply joshix's suggestions to Readme

colhom commented:
@joshix suggestions digested

joshix commented on README.md:
```diff
@@ -0,0 +1,79 @@
+
+A khealth *Module* is a single command that invokes a set of Routines and a single Collector. The Collector gathers events from the Routines and exposes metrics on its status endpoint.
+
+* **cmd/**:
```
```
**`cmd/`**
```

 (mono font for file path)

joshix commented on README.md:
```diff
@@ -0,0 +1,79 @@
+
+* **cmd/**:
+
+  This is where the `Module` entrypoint programs live. Each `Module` should have exactly one `main` package package in an eponymous directory beneath `cmd/`.
```
"package" is repeated. Delete one.

joshix commented on README.md:
```diff
@@ -0,0 +1,79 @@
+
+  This is where the `Module` entrypoint programs live. Each `Module` should have exactly one `main` package package in an eponymous directory beneath `cmd/`.
+
+* **pkg/routines**:
```
```
**`pkg/routines/`**
```

(mono, bold. file path)

joshix commented on README.md:
````diff
@@ -0,0 +1,79 @@
+	Cleanup() error
+  }
+  ```
+  `Init` is called, and then `Poll` in a loop. Once the ttl has expired, `Cleanup` is called. Rinse-repeat. Every step generates an event which is received by the collector.
````
I liked the informality, but:
`Init` is called, and then `Poll` in a loop. When the TTL expires, `Poll` terminates, and `Cleanup` is called. Each iteration of this cycle generates an event and sends it to the Collector.

joshix commented on README.md:
````diff
@@ -0,0 +1,79 @@
+  ```
+  `Init` is called, and then `Poll` in a loop. Once the ttl has expired, `Cleanup` is called. Rinse-repeat. Every step generates an event which is received by the collector.
+
+  The `NewRoutine` function can be used to create a new Routine.
````
The `NewRoutine` function [takes x, y, z where x/y/z may be args(?) and] returns a [pointer to?] a new khealth Routine.

joshix commented on README.md:
```diff
@@ -0,0 +1,79 @@
+
+  The `NewRoutine` function can be used to create a new Routine.
+
+* **pkg/collectors**:
```
```
**`pkg/collectors/`** 
```

(monospace literal string of relative file path)

(also, I'm adding the trailing slash you used with `cmd/` to other rel paths in these bold points)

joshix commented on README.md:
````diff
@@ -0,0 +1,79 @@
+	Terminate() error
+  }
+  ```
+  Collectors should implement the Collector interface and make use of routines. The pattern for wiring up routines inside a collector implementation is as follows:
````
s/should/must/ && s/routines/Routines/g && s/collector/Collector/

"To wire Routines to a Collector implementation, follow this general pattern:" maybe cleaner phrasing?

joshix commented on README.md:
```diff
@@ -0,0 +1,79 @@
+      `SignalTerminate` non-blocking, so afterwards you'll want to block until each Routines' `Events` channel has emitted a `nil` value.
+
+## Modules
+- **cmd/rscheduler**
```
```
**`cmd/rcscheduler`**
```

(missing `c`, use monospace)

joshix commented on README.md:
```diff
@@ -0,0 +1,79 @@
+## Roadmap
+- More routines: We want routines that do everything! Test network latency. Write to disk. Compute fibonacci sequences.
+
+* Prometheus integration: Collectors expose prometheus-compatbile status endpoints and metrics. Pre-built infrastructure for aggregating metrics from a dynamic set of canary pods purpose-built to exercise Kubernetes clusters in a specific way.
```
Prometheus integration: Collectors expose Prometheus-compatible status endpoints and metrics, providing readymade infrastructure to aggregate statistics from a set of canary pods, designed specifically to exercise Kubernetes cluster resources.

(Needed to be a complete sentence, and easiest way here is making it a really long one.)

joshix commented on README.md:
```diff
@@ -0,0 +1,79 @@
+
+* Prometheus integration: Collectors expose prometheus-compatbile status endpoints and metrics. Pre-built infrastructure for aggregating metrics from a dynamic set of canary pods purpose-built to exercise Kubernetes clusters in a specific way.
+
+* Alerting: Use experimental [alertmanager](https://github.com/prometheus/alertmanager) to alert on metrics
```
s/Use/Use the/

joshix commented on README.md:
```diff
@@ -0,0 +1,79 @@
+
+## Who should use this?
+
+* **Cluster administrators**: Gain insight into your kubernetes cluster's performance.  Monitor health endpoints which report on various testing routines.
```
Kubernetes (whereas "routines", being used in the more general sense, is ok in lower case here)

joshix commented:
@colhom sorry for that misfire and the comment churn. Comments now inline to tip of this PR. LGTM after those fixes.

colhom committed:
docs: further clarification, formatting fixes

consistent text formatting for recurring entities
clearer english, proper grammar
explain NewRoutine method in full
